cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -41,42 +68,97 @@
                              StoreFactory storeFactory, ProgramRuntimeService programRuntimeService,
                              PreferencesStore preferencesStore) {
+    this.pollingDelay = cConf.getLong(Constants.Notification.Stream.STREAM_SIZE_SCHEDULE_POLLING_DELAY);
     this.notificationService = notificationService;
     this.streamAdmin = streamAdmin;
     this.storeFactory = storeFactory;
     this.programRuntimeService = programRuntimeService;
     this.preferencesStore = preferencesStore;
+    this.streamSubscribers = Maps.newConcurrentMap();
+    this.scheduleSubscribers = Maps.newConcurrentMap();
+    this.store = null;
   }
 
   public void start() {
-
+    notificationExecutor = Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("stream-size-scheduler-%d"));
+    streamPollingExecutor = Executors.newScheduledThreadPool(STREAM_POLLING_THREAD_POOL_SIZE,
+                                                             Threads.createDaemonThreadFactory("stream-polling-%d"));
   }
 
   public void stop() {
-
+    streamPollingExecutor.shutdownNow();
+    for (StreamSubscriber subscriber : streamSubscribers.values()) {
+      subscriber.cancel();
+    }
   }
 
   @Override
   public void schedule(Id.Program program, SchedulableProgramType programType, Schedule schedule) {
     Preconditions.checkArgument(schedule instanceof StreamSizeSchedule,
                                 "Schedule should be of type StreamSizeSchedule");
+    StreamSizeSchedule streamSizeSchedule = (StreamSizeSchedule) schedule;
+    schedule(program, programType, streamSizeSchedule, true, -1, -1, true);
+  }
+
+  private void schedule(Id.Program program, SchedulableProgramType programType, StreamSizeSchedule streamSizeSchedule,
+                        boolean active, long baseRunSize, long baseRunTs, boolean persist) {
+    try {
+      // Create a new StreamSubscriber, if one doesn't exist for the stream passed in the schedule
+      Id.Stream streamId = Id.Stream.from(program.getNamespaceId(), streamSizeSchedule.getStreamName());
+      StreamSubscriber streamSubscriber = new StreamSubscriber(streamId);
+      synchronized (this) {
+        // This block is synchronized so that we can't have the following situation:
+        // - creation of a schedule, using an existing StreamSubscriber which has one other schedule
+        // - deletion of the other schedule, leading to deletion of existing StreamSubscriber
+        // - add the the first schedule to the StreamSubscriber, which has been removed from streamSubscribers
+
+        StreamSubscriber previous = streamSubscribers.putIfAbsent(streamId, streamSubscriber);
+        if (previous == null) {
+          streamSubscriber.start();
+        } else {
+          streamSubscriber = previous;
+        }
+
+        // Add the scheduleTask to the StreamSubscriber
+        streamSubscriber.createScheduleTask(program, programType, streamSizeSchedule,
```

Comment:
Then the ``streamSubscriber.createScheduleTask`` method won't replace the existing schedule. I should make this method return ``true`` when we do add a new schedule, so that the next line, 
```
scheduleSubscribers.put(getScheduleId(program, programType, streamSizeSchedule.getName()), streamSubscriber);
```
is executed only if the call returned ``true``.